File: RecordTechnology/RecordTechnology/201711/HXAppPod/Utils/HXSDirectoryManager.m
//
//  HXSDirectoryManager.m
//  store
//
//  Created by Arthur Wang on 14/10/26.
//  Copyright (c) 2014年 Arthur Wang All rights reserved.
//

#import "HXSDirectoryManager.h"
#import <sys/xattr.h>
#import "NSString+Addition.h"

@implementation HXSDirectoryManager

+ (NSString *) getLibraryDirectory
{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    
    return documentsDirectory;
}

+ (NSString *) getCachesDirectory
{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSMutableString *path = [paths objectAtIndex:0];
    
    return path;
}

+ (BOOL) createDirectoryAtCachesWithName:(NSString *)dirName
{
    NSString *dirPath = [[HXSDirectoryManager getCachesDirectory] stringByAppendingPathComponent:dirName];
    
    return [HXSDirectoryManager createDirectoryAtPath:dirPath];
}

+ (NSString *) getDocumentsDirectory
{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    
    return documentsDirectory;
}

+ (NSArray*) enumarateDirectoryAtPath:(NSString *)path inCaches:(BOOL)isInCaches
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSString *enumPath;
    NSArray *enumArray;
    
    if (isInCaches)
        enumPath = [NSString stringWithString:[[HXSDirectoryManager getCachesDirectory] stringByAppendingPathComponent:path]];
    else
        enumPath = [NSString stringWithString:path];
    
    
    enumArray = [fileManager contentsOfDirectoryAtPath:enumPath error:nil];
    
    return enumArray;
}

+ (NSString *) getTempDirectory
{
    return NSTemporaryDirectory();
}

+ (BOOL) createDirectoryAtPath:(NSString *)path
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if ([fileManager fileExistsAtPath:path isDirectory:nil])
    {
        return NO;
    }
    else
    {
        return [fileManager createDirectoryAtPath:path withIntermediateDirectories:NO attributes:NULL error:NULL];
    }
    
}

//+ (NSArray*) enumarateDirectoryAtPath:(NSString *)path inDocuments:(BOOL)isInDocuments {
//    NSFileManager *fileManager = [NSFileManager defaultManager];
//    NSString *enumPath;
//    NSArray *enumArray;
//
//    if (isInDocuments)
//        enumPath = [NSString stringWithString:[[DirectoryManager getDocumentsDirectory] stringByAppendingPathComponent:path]];
//    else
//        enumPath = [NSString stringWithString:path];
//
//
//    enumArray = [fileManager contentsOfDirectoryAtPath:enumPath error:nil];
//
//    return enumArray;
//}

//+ (BOOL) createDirectoryAtDocumentsWithName:(NSString *)dirName {
//    NSString *dirPath = [[DirectoryManager getDocumentsDirectory] stringByAppendingPathComponent:dirName];
//
//    return [DirectoryManager createDirectoryAtPath:dirPath];
//}

//+ (NSMutableArray*)createURLStartWith:(NSString *)start AtPath:(NSString *)path inDocuments:(BOOL)isInDocuments {
//    NSMutableArray *result = [[NSMutableArray alloc] init];
//    NSArray *fileList = [DirectoryManager enumarateDirectoryAtPath:path inDocuments:isInDocuments];
//
//    if (path != nil) {
//        for (int i = 0; i < fileList.count; ++i) {
//            NSString *tmpString = [[NSString alloc] initWithFormat:@"%@%@/%@", start, path, [fileList objectAtIndex:i]];
//            [result addObject:tmpString];
//            [tmpString release];
//        }
//    } else {
//        for (int i = 0; i < fileList.count; ++i) {
//            NSString *tmpString = [[NSString alloc] initWithFormat:@"%@%@", start, [fileList objectAtIndex:i]];
//            [result addObject:tmpString];
//            [tmpString release];
//        }
//    }
//    return result;
//}

+ (void)moveFilesFrom:(NSString *)fromFile to:(NSString *)toFile
{
    NSError * error;
    NSArray * files = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:fromFile error:&error];
    
    if(files && files.count > 0)
    {
        for(NSString * selectedFile in files)
        {
            NSString * fromFilePath = [fromFile stringByAppendingPathComponent: selectedFile];
            NSString * toFilePath = [toFile stringByAppendingPathComponent: selectedFile];
            
            BOOL isDir;
            //判断是否是为目录
            
            if ([[NSFileManager defaultManager] fileExistsAtPath:fromFilePath isDirectory:&isDir] && isDir)
            {//目录
                
                [[NSFileManager defaultManager] createDirectoryAtPath:toFilePath withIntermediateDirectories:NO attributes:nil error:nil];
                
                [HXSDirectoryManager moveFilesFrom:fromFilePath to:toFilePath];
            }
            else
            {//文件
                [[NSFileManager defaultManager] moveItemAtPath:fromFilePath toPath:toFilePath error:nil];
            }
        }
    }
}

+ (BOOL)addSkipBackupAttributeToItemAtURL:(NSURL *)URL
{
    if(URL == nil)
        return NO;
    assert([[NSFileManager defaultManager] fileExistsAtPath: [URL path]]);
    
    NSError *error = nil;
    [URL setResourceValue:[NSNumber numberWithBool:YES] forKey:NSURLIsExcludedFromBackupKey error:&error];

    return error == nil;
}

+ (NSString *)getAdImagePath:(NSString *)image_url {
    NSString *imagePath = [HXSDirectoryManager getCachesDirectory];
    imagePath = [imagePath stringByAppendingPathComponent:@"LaunchAd"];
    if (![[NSFileManager defaultManager] fileExistsAtPath:imagePath isDirectory:nil])
    {
        [HXSDirectoryManager createDirectoryAtPath:imagePath];
    }
    imagePath = [imagePath stringByAppendingString:[NSString md5:image_url]];
    return imagePath;
}

@end

File: RecordTechnology/RecordTechnology/201711/HXAppPod/Utils/HXSDirectoryManager.swift
//
//  HXSDirectoryManager.swift
//  RecordTechnology
//
//  Created by ArthurWang on 2017/11/27.
//  Copyright © 2017年 Monkey King. All rights reserved.
//

import Foundation

class HXSDirectoryManager: NSObject {
    @objc class func getLibraryDirectory() -> String {
        let paths = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.libraryDirectory, FileManager.SearchPathDomainMask.userDomainMask, true);
        let libraryDirectory = paths[0];

        return libraryDirectory;
    }

    @objc class func getCachesDirectory() -> String {
        let paths = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.cachesDirectory, FileManager.SearchPathDomainMask.userDomainMask, true);
        let cachesDirectory = paths[0];

        return cachesDirectory;
    }

    @objc class func getDocumentsDirectory() -> String {
        let paths = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.documentDirectory, FileManager.SearchPathDomainMask.userDomainMask, true);
        let documentsDirectory = paths[0];

        return documentsDirectory;
    }

    @objc class func getTempDirectory() -> String {
        return NSTemporaryDirectory();
    }





    @objc class func enumarateDirectoryAtPath(path:String, inCaches:Bool) -> NSArray? {
        var enumPath:String? = nil;

        if inCaches {
            let cachesDir = self.getCachesDirectory() as NSString;
            enumPath = cachesDir.appendingPathComponent(path);
        } else {
            enumPath = String.init(path);
        }

        if nil == enumPath {
            return nil;
        }

        do {
            let enumArray = try FileManager.default.contentsOfDirectory(atPath: enumPath!);

            return enumArray as NSArray;
        } catch {
            return nil;
        }
    }

    @objc class func createDirectoryAtPath(path:String) -> Bool {
        if FileManager.default.fileExists(atPath: path) {
            return false;
        } else {
            do {
                try FileManager.default.createDirectory(atPath: path, withIntermediateDirectories: false, attributes: nil);
            } catch {
                return false;
            }

            return true;
        }
    }

    @objc class func createDirectoryAtCachesWithName(dirName:String) -> Bool {
        let path = self.getCachesDirectory() as NSString;
        let dirPath = path.appendingPathComponent(dirName);

        return self.createDirectoryAtPath(path: dirPath);
    }

    @objc class func moveFilesFrom(fromFile:String, toFile:String) {
        do {
            let files = try FileManager.default.contentsOfDirectory(atPath: fromFile);

            if 0 < files.count {
                for selectedFile in files {
                    let fromFilePath = (fromFile as NSString).appendingPathComponent(selectedFile);
                    let toFilePath = (toFile as NSString).appendingPathComponent(selectedFile);

                    let isDir = UnsafeMutablePointer<ObjCBool>.allocate(capacity: 1);
                    if FileManager.default.fileExists(atPath: fromFilePath, isDirectory:isDir) {
                        if isDir.pointee.boolValue {
                            do {
                                try FileManager.default.createDirectory(atPath: toFilePath, withIntermediateDirectories: false, attributes: nil);
                                self.moveFilesFrom(fromFile: fromFilePath, toFile: toFilePath);
                            } catch {
                                // Do nothing
                            }
                        } else {
                            do {
                                try FileManager.default.moveItem(atPath: fromFilePath, toPath: toFilePath);
                            } catch {
                                // Do nothing
                            }
                        }
                    }
                }
            }
        } catch {
            // Do nothing
        }
    }

    @objc class func addSkipBackupAttributeToItemAtURL(URL:NSURL?) -> Bool {
        if nil == URL {
            return false;
        }

        assert(FileManager.default.fileExists(atPath: (URL?.path)!));

        do {
            try URL?.setResourceValue(NSNumber.init(value: true), forKey: URLResourceKey.isExcludedFromBackupKey);

            return true;
        } catch {
            return false;
        }
    }

    @objc class func getAdImagePath(image_url:String) -> String {
        var imagePath = self.getCachesDirectory();
        imagePath = (imagePath as NSString).appendingPathComponent("LaunchAd");
        if !(FileManager.default.fileExists(atPath: imagePath, isDirectory: nil)) {
            _ = self.createDirectoryAtPath(path: imagePath);
        }

        imagePath = (imagePath as NSString).appending(NSString.md5(image_url));

        return imagePath;
    }

}







